app/main.py

```diff
@@ -1,214 +1,49 @@
-
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
-
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
-
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
-
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
-
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
-
-    # arthur
-    arthur = knightsConfig["arthur"]
-
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
-
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
-
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
-
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
-
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
-
-    # mordred
-    mordred = knightsConfig["mordred"]
-
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
-
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
+from app.Battle.Fight import fight
+from app.knights.Create_Knight import Knight
+
+
+def battle(knights: dict) -> dict:
+    lancelot = Knight(
+        knights["lancelot"]["name"],
+        knights["lancelot"]["power"],
+        knights["lancelot"]["hp"]
+    )
+    arthur = Knight(
+        knights["arthur"]["name"],
+        knights["arthur"]["power"],
+        knights["arthur"]["hp"]
+    )
+    mordred = Knight(
+        knights["mordred"]["name"],
+        knights["mordred"]["power"],
+        knights["mordred"]["hp"]
+    )
+    red_knight = Knight(
+        knights["red_knight"]["name"],
+        knights["red_knight"]["power"],
+        knights["red_knight"]["hp"]
+    )
+
+    list_of_knights = [lancelot, arthur, mordred, red_knight]
+    list_of_actions = ["weapon", "armour", "potion"]
+
+    for action in list_of_actions:
+        for knight in list_of_knights:
+            knight_name = knight.name.lower().replace(" ", "_")
+            if action == "weapon":
+                knight.apply_weapon(knights[knight_name][action])
+            elif action == "armour":
+                knight.apply_armour(knights[knight_name][action])
+            elif action == "potion":
+                if knights[knight_name]["potion"] is not None:
+                    knight.apply_potion(knights[knight_name][action])
```

Comment:
Move this part of the code to a separate function because it is a separate logical operation.